pkg/services/ngalert/notifier/multiorg_alertmanager.go

```diff
@@ -206,6 +170,62 @@ func (moa *MultiOrgAlertmanager) SyncAlertmanagersForOrgs(ctx context.Context, o
 	}
 }
 
+// addOrUpdateAlertmanagerConfigurations retrieves the latest configurations for all organizations and then calls addOrUpdateAlertmanager for every organization in orgIDs.
+//The result is either a map that contains ID of all organizations updated or created or error in the case something happened while fetching configurations from database
+//NOTE this is an internal method, and it is not supposed to be called without locked mutex alertmanagersMtx
+func (moa *MultiOrgAlertmanager) addOrUpdateAlertmanagerConfigurations(ctx context.Context, orgIDs []int64) (map[int64]struct{}, error) {
+	orgsFound := make(map[int64]struct{}, len(orgIDs))
+	dbConfigs, err := moa.getLatestConfigs(ctx)
```

Comment:
I don't think the order of operations in this refactor is correct wrt to the lock. You take the lock before executing a database call and in the previous version we didn't do that deliberately for two reasons:

1. The database call has the potential to be quite slow even for simple queries (network, load, etc) - which would end up taking lock for longer than necessary and block API calls.
2. Because the database call is the only thing that can fail, you risk taking the lock unnecessarily.

Reply:
Agree. I will move it back.